homeassistant/components/knx/light.py

```diff
@@ -163,10 +163,15 @@ async def async_turn_on(self, **kwargs: Any) -> None:
         white_value = kwargs.get(ATTR_WHITE_VALUE, self.white_value)
         mireds = kwargs.get(ATTR_COLOR_TEMP, self.color_temp)
 
-        update_brightness = ATTR_BRIGHTNESS in kwargs
-        update_color = ATTR_HS_COLOR in kwargs
-        update_white_value = ATTR_WHITE_VALUE in kwargs
-        update_color_temp = ATTR_COLOR_TEMP in kwargs
+        update_brightness = self._device.supports_brightness and brightness is not None
+        update_color = (
+            self._device.supports_rgbw or self._device.supports_color
+        ) and hs_color is not None
+        update_white_value = self._device.supports_rgbw and white_value is not None
+        update_color_temp = (
+            self._device.supports_color_temperature
+            or self._device.supports_tunable_white
+        ) and mireds is not None
```

Comment:
This would mean every call to turn_on would trigger telegrams for each supported feature. This is not desirable as it may conflict with internal actuator settings. 

E.g., When a light supporting brightness and color_temp gets a call `async_turn_on(brightness=60)` it would send to its brightness address (the new value) and its color_temp address (its current value - see line 164).

 It shall just send the values the user intended to change.

Reply:
Sorry, I've overseen the default ...

Added a test to check that wrong behavior and fixed it.

Do you mind that double dictionary access?
```
 brightness = kwargs.get(ATTR_BRIGHTNESS, self.brightness)

...

update_brightness = (
            self._device.supports_brightness and ATTR_BRIGHTNESS in kwargs
        )
```
I could introduce:
```
    def _determine(self, is_supported: bool, new_value, old_value) -> list[Any, bool]:
        if not is_supported:
            return [None, False]
        if new_value is not None:
            return [new_value, True]
        return [old_value, False]
```
then it would be:
```
        brightness, update_brightness = self._determine(
            self._device.supports_brightness,
            kwargs.get(ATTR_BRIGHTNESS),
            self.brightness,
        )
```
with only a single dictionary access.

Reply:
I don't really mind. If you really like to save dict lookups you could probably remove the update_* and just lookup the ATTR* if supported
```python
# -- mireds = kwargs.get(ATTR_COLOR_TEMP, self.color_temp)
mireds = kwargs.get(ATTR_COLOR_TEMP) if self._device.supports_color_temperature or self._device.supports_tunable_white else None

# -- update_color_temp = (
# --             self._device.supports_color_temperature
# --             or self._device.supports_tunable_white
# --         ) and ATTR_COLOR_TEMP in kwargs

# -- if update_color_temp:
if mireds is not None:
```
then retrieve the default values only if needed inside the `if` blocks.

Reply:
I have added some more tests. Simplified test result checking by extending the mock and reduced the dict accesses.